File: 303_Binary_Sum_Query_Immutable/sumRange_v2.py
class NumArray(object):
    def __init__(self, nums):
        """
        initialize your data structure here.
        :type nums: List[int]
        """
        self.size = len(nums)
        self.BITTree = [0] * (len(nums) + 1)
        for i in xrange(self.size):
            self.updatebit(i, nums[i])

    def updatebit(self, i, val):
        i += 1
        while i <= self.size:
            self.BITTree[i] += val
            i += i & (-i)

    def getSum(self, i):
        sum = 0
        i += 1
        while i > 0:
            sum += self.BITTree[i]
            i -= i & (-i)
        return sum

    def sumRange(self, i, j):
        """
        sum of elements nums[i..j], inclusive.
        :type i: int
        :type j: int
        :rtype: int
        """
        return self.getSum(j) - self.getSum(i - 1)

nums = [-2, 0, 3, -5, 2, -1]
na = NumArray(nums)
print na.sumRange(0, 5)

File: 303_Binary_Sum_Query_Immutable/sumRange_v2.cpp
struct SegmentTreeNode {
    int start, end, sum;
    SegmentTreeNode* left;
    SegmentTreeNode* right;
    SegmentTreeNode(int a, int b): start(a), end(b), left(NULL), right(NULL) {}
};

class NumArray {
public:
    NumArray(vector<int> &nums) {
        int n = nums.size();
        root = buildTree(nums, 0, n - 1);
    }

    int sumRange(int i, int j) {
        return queryTree(i, j, root);
    }

    SegmentTreeNode* buildTree(vector<int>& nums, int start, int end) {
        if(start > end) { return NULL; }
        SegmentTreeNode* root = new SegmentTreeNode(start, end);

        if(start == end) {
            root->sum = nums[start];
            return root;
        }

        int mid = start + (end - start) / 2;

        root->left = buildTree(nums, start, mid);
        root->right = buildTree(nums, mid + 1, end);
        root->sum = root->left->sum + root->right->sum;
        return root;
    }

    int queryTree(int i, int j, SegmentTreeNode* root) {
        if(!root) { return 0; }
        if(i == root->start && j == root->end) { return root->sum; }
        int mid = (root->start + root->end) / 2;
        if(j <= mid) { return queryTree(i, j, root->left); }
        if(i > mid) { return queryTree(i, j, root->right); }
        return queryTree(i, mid, root->left) + queryTree(mid + 1, j, root->right);
    }

private:
    SegmentTreeNode* root;
};
